tests/unit/adaptors/graph/test_hierarchy.py

```diff
@@ -115,3 +115,22 @@ def test_find_descendants_works_if_modules_added_in_different_order():
         "mypackage.foo.blue.alpha",
         "mypackage.foo.blue.alpha.one",
     }
+
+
+class TestFindMatchingModules:
+    def test_finds_matching_modules(self):
+        graph = ImportGraph()
+        graph.add_module("foo")
+        graph.add_module("foo.bar")
+        graph.add_module("foo.bar.baz")
+
+        assert graph.find_matching_modules("foo") == {"foo"}
+        assert graph.find_matching_modules("foo.*") == {"foo.bar"}
+        assert graph.find_matching_modules("foo.**") == {"foo.bar", "foo.bar.baz"}
```

Comment:
🍷 This test basically duplicates the rust test. For me it's okay to keep both.

Reply:
I think a good rule of thumb is that the Python tests should be comprehensive, and if we want to add some Rust tests to help too (even if they duplicate) then that's fine. Do you agree?

Reply:
> Do you agree?

I think I could be persuaded either way. In the sense of a testing pyramid, one might argue that the rust tests should be the most comprehensive, since this is the lowest layer of the code. All the python tests integrate the rust code, so by definition are higher up the pyramid, so we'd expect less comprehensive tests there. 

That said, this isn't a normal testing pyramid, and the preexisting code puts most of the tests in python. 

In summary I think it's good to have both, certainly with comprehensive python tests. I'll flesh out the python tests a bit more.

Reply:
Thinking back to when we swapped in your Rust implementation, it was a really useful property of the test suite that  there was thorough test coverage of the Python API - we just deleted all the Rust tests and still felt confident. If coverage had been split across the two languages then we wouldn't have been able to do that.

So I think focusing on the Python tests is worthwhile - we can treat the Rust tests more lightly as tests to help us as we go along, but more disposable.

Arguably the Python unit tests are still at the base of the testing pyramid - they're fast and give us lots of signal about granular behaviour.

Reply:
Yes okay I think you're right 👍